feat(major): transform add url to option and document more possible use cases

samzilverberg opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

resolves issue https://github.com/fastify/fastify-swagger/issues/555

combines the following PRs into one by enhancing the existing `transform` option:

https://github.com/fastify/fastify-swagger/pull/559
https://github.com/fastify/fastify-swagger/pull/560

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

mcollina reviewed (approved):
lgtm

samzilverberg commented on README.md:
```diff
@@ -225,7 +225,7 @@ An example of using `fastify-swagger` with `static` mode enabled can be found [h
  | stripBasePath      | true             | Strips base path from routes in docs.                                                                                     |
  | swagger            | {}               | [Swagger configuration](https://swagger.io/specification/v2/#swaggerObject).                                              |
- | transform          | null             | Transform method for schema.                                                                                              |
+ | transform          | null             | Transform method for the route's schema and url: `(schema: any, url: string) => { schema: any, url: string }`.                                                                                              |
```
**IMPORTANT: this is a breaking change that requires a major version release**

it changes the function return value.
so any existing implementations will have to be migrated from returning simple `return transformed` to `return { schema: transformed, url }`

samzilverberg replied:
we can add some fallback to support the old code (and release as minor), but it'l be a little ugly.
fallback code is usually ugly  😆

climba03003 replied:
I would use a single object for the parameter.
`{ schema, url } => { schema, url }`

samzilverberg replied:
done, but now its even a bigger breaking change :) need to release a BIG major

climba03003 replied:
It is OK, the current syntax is much extendable in future.

climba03003 commented on lib/spec/openapi/index.js:
```diff
@@ -26,18 +26,20 @@ module.exports = function (opts, cache, routes, Ref, done) {
       if (shouldRouteHide(schema, shouldRouteHideOpts)) continue
 
-      const url = normalizeUrl(route.url, defOpts.servers, defOpts.stripBasePath)
+      let url = transformResult.url ?? route.url
```
Nullish operator add on `node@14` and we support `node@10` currently.
```suggestion
      let url = transformResult.url || route.url
```

climba03003 commented on lib/spec/swagger/index.js:
```diff
@@ -26,18 +26,20 @@ module.exports = function (opts, cache, routes, Ref, done) {
+        ? defOpts.transform(route.schema, route.url)
+        : {}
 
+      const schema = transformResult.schema ?? route.schema
```
```suggestion
      const schema = transformResult.schema || route.schema
```

climba03003 commented on index.d.ts:
```diff
@@ -127,9 +127,9 @@ export interface FastifyDynamicSwaggerOptions extends FastifySwaggerOptions {
+   * custom function to transform the route's schema and url
    */
-  transform?: Function;
+  transform?: ({schema, url}: {schema: JSONObject, url: string}) => { schema: JSONObject, url: string };
```
The input is not restricted to JSON object. It can be anything as it is input from user.
```suggestion
  transform?: <S = any>({schema, url}: {schema: S, url: string}) => { schema: JSONObject, url: string };
```

samzilverberg replied:
done, I didn't commit ur suggestion directly because `JSONObject` was still on the response

zekth reviewed (approved):
lgtm

samzilverberg commented:
rebased on latest main branch and squashed. if you don't have any more suggestions feel free to merge.

JUST REMEMBER this introduces breaking changes so better release as major (or maybe minor will do too, just not patch.......)

darkgl0w commented:
> rebased on latest main branch and squashed. if you don't have any more suggestions feel free to merge.
> 
> JUST REMEMBER this introduces breaking changes so better release as major (or maybe minor will do too, just not patch.......)

It's because the returned schema is a `JSONObject` ^^

darkgl0w commented on README.md:
```diff
@@ -225,7 +225,7 @@ An example of using `fastify-swagger` with `static` mode enabled can be found [h
  | stripBasePath      | true             | Strips base path from routes in docs.                                                                                     |
  | swagger            | {}               | [Swagger configuration](https://swagger.io/specification/v2/#swaggerObject).                                              |
- | transform          | null             | Transform method for schema.                                                                                              |
+ | transform          | null             | Transform method for the route's schema and url: `({schema: JSONObject, url: string}) => { schema: JSONObject, url: string }`.                                                                                              |
```
This needs to be changed according to @climba03003 suggestion.

darkgl0w commented on index.d.ts:
```diff
@@ -127,9 +127,9 @@ export interface FastifyDynamicSwaggerOptions extends FastifySwaggerOptions {
+   * custom function to transform the route's schema and url
    */
-  transform?: Function;
+  transform?: <S = any>({schema, url}: {schema: S, url: string}) => { schema: S, url: string };
```
```suggestion
  transform?: <S = any>({schema, url}: {schema: S, url: string}) => { schema: JSONObject, url: string };
```

@climba03003 I would go with `unknown` instead of `any` for a user input WDYT ?

climba03003 replied:
My intention here is not forcing the user to specify a type and they are good to go. If it is `= unknown`, that means the user is force to specify the type, then I would go for remove the `= unknown`.

samzilverberg replied:
S could be `FastifySchema` (by either direct ref or copy paste the type into this lib) :

https://github.com/fastify/fastify/blob/596e7cad346fa31171f26260a9612b20f4646bcc/types/schema.d.ts#L9-L15

is it a good fit?

climba03003 replied:
Yes, it is a possible change. 
```
<S extends FastifySchema = FastifySchema>({schema, url}: {schema: S, url: string}) => { schema: JSONObject, url: string }
```

darkgl0w replied:
This one is better IMO. ^^

samzilverberg replied:
done and it passes the tests 🥳 

I also squashed the commits and force pushed again to make history a lil nicer.

Eomm commented on README.md:
```diff
@@ -241,28 +241,49 @@ If you set `exposeRoute` to `true` the plugin will expose the documentation with
-To use different schemas such as [Joi](https://github.com/hapijs/joi) you can pass a synchronous `transform` method in the options to convert them back to standard JSON schemas expected by this plugin to generate the documentation (`dynamic` mode only).
+By passing a synchronous `transform` function you can modify the route's url and schema.
+
+some possible uses of this are:
```
```suggestion
Some possible uses of this are:
```

Eomm commented on README.md:
```diff
@@ -241,28 +241,49 @@ If you set `exposeRoute` to `true` the plugin will expose the documentation with
+- altering the route url into something that's more suitable for the api spec
+- using different schemas such as [Joi](https://github.com/hapijs/joi) and transforming them to standard JSON schemas expected by this plugin
+
+this option is available in `dynamic` mode only.
```
```suggestion
This option is available in `dynamic` mode only.
```

Eomm commented on README.md:
```diff
@@ -241,28 +241,49 @@ If you set `exposeRoute` to `true` the plugin will expose the documentation with
+
+this option is available in `dynamic` mode only.
+
+example of all the possible uses mentioned:
```
```suggestion
Examples of all the possible uses mentioned:
```

Eomm reviewed (approved):
Just some nit

darkgl0w commented on README.md:
```diff
@@ -241,30 +241,54 @@ If you set `exposeRoute` to `true` the plugin will expose the documentation with
-    if (body) transformed.body = convert(body)
-    if (querystring) transformed.querystring = convert(querystring)
-    return transformed
+    const transformedSchema = { ...rest }
```
```suggestion
    const transformedSchema = Object.assign({}, rest)
```
Replacing rest by something better suited (see my previous comment).

darkgl0w replied:
> why the Object.assign on target empty object? which is essesntialy a shallow copy over if i understand it correctly.

Here `Object.assign({}, rest)` is equivalent to `{ ...rest }` (it is just more explicit) ^^

> we're not mutating anything anywhere, should be ok to just use the original ...rest, or i'm missing something?

By essence `transform` will mutate the schema (and it's what it does in this example), this is why we initialize a new object instead of just doing something like this:

```js
// ...
const { params, body, querystring, headers, response, ...transformedSchema } = schema
let transformedUrl = url

if (params) transformedSchema.params = convert(params)
// ...
```

samzilverberg replied:
hmm, this makes me think of 2 things:

1) if we want to ensure no mutation of existing `route.schema` which was passed in, then probably it'd be better to pass `lodash.cloneDeep(route.schema)` to the transform fn.

2)  returned schema should maybe also be some subtype of fastify schema.
to avoid users returning just part of the schema by mistake (without the `rest`).
```
transform?: <S extends FastifySchema = FastifySchema, T extends FastifySchema = FastifySchema>({schema, url}: {schema: S, url: string}) => { schema: T, url: string };
```

WDYT?

darkgl0w replied:
No need to add a new dependency.
It's up to the user to decide if he wants to mutate the original schema or not (`transform` is about letting users apply the mutations of their choice after all 😄). In the documentation we just need to promote good practices IMO and by creating a new object with either `{ ...rest }` or `Object.assign({}, rest)` we do so and it is enough. 😛 

And about the second part, we expect a json schema as the output so checking that it is a json Object is enough and once again it's up to the user to make sure his transform function works as it should 😜

darkgl0w commented on README.md:
```diff
@@ -241,30 +241,54 @@ If you set `exposeRoute` to `true` the plugin will expose the documentation with
+  transform: ({ schema, url }) => {
     const {
       params = undefined,
       body = undefined,
       querystring = undefined,
-      ...others
+      headers = undefined,
+      response = undefined,
+      ...rest
```
No need to assign undefined as default fallback values here, it is automatically done when using object destructuring.

darkgl0w reviewed (commented):
A few little nitpicks.

Other than that it seems good to me :)

climba03003 reviewed (approved):
LGTM.
